File: C/Euler03/Euler03.c
// C NELSON 2020
/* The prime factors of 13195 are 5, 7, 13 and 29.
What is the largest prime factor of the number 600851475143
(Answer = 6857) */

// solution divides out all factors until largest prime factor remains

#include <stdio.h>
#include <time.h>
#include <math.h>

int main()
{
    long long finalAnswer = 0;
    long long n = 600851475143;
    int div = 3;
    
    clock_t tic, toc;
    double timeElapsed;
    
    tic = clock();
    
    int nsqrt = sqrt(n);

    if ((n & 1) != 1)
    {
        while ((n & 1) != 1)
        {
            n = n / 2;
        }
        if (n == 1)
        {
            div = 2;
        }
    }

    while (n > 1 && div < nsqrt)
    {
        if (n % div == 0)
        {
            n = n / div;
        }
        else
        {
            div += 2;
        }
    }

    finalAnswer = (n != 1) ? n : div;

    toc = clock();
    timeElapsed = (double)(toc - tic) / CLOCKS_PER_SEC;
    
    printf("Answer:\t %lld\n", finalAnswer);
    printf("Took:\t %.20f\n", timeElapsed);

    return 0;
}

File: Lua/Euler03/Euler03.lua
-- C NELSON 2021
--[[ The prime factors of 13195 are 5, 7, 13 and 29.
What is the largest prime factor of the number 600851475143
(Answer = 6857) --]]

-- solution divides out all factors until largest prime factor remains

finalanswer = 0
n = 600851475143
div = 3

tic = os.clock()

nsqrt = math.sqrt(n)

if (n % 2 == 0) then
    while ( n % 2 == 0 ) do
        n = n / 2
    end
    if (n == 1) then
        div = 2
    end
end

while (n > 1 and div < nsqrt) do
    if (n % div == 0) then
        n = n / div
    else
        div = div + 2
    end
end

if (n ~= 1) then
    finalanswer = n
else
    finalanswer = div
end

toc = os.clock()
timeElapsed = toc - tic

print(string.format("Answer:\t%s", finalanswer))
print(string.format("Took:\t%.6f seconds", timeElapsed))